execute/README.md

```diff
@@ -0,0 +1,458 @@
+|FlushKey|group key|Data associated with the given group key should be flushed from the dataset
+|Finish|error (optional)|The upstream dataset will produce no more messages
+
+The following messages exist, but are deprecated and should not be used by future transformations.
+
+|Name|Data|Description
+|----|----|-----------
+|Process|flux.Table|A full table is ready to be processed
+|UpdateWatermark|time|Data older than the given time will not be sent
+|UpdateProcessingTime|time|Marks the present time
+|RetractTable|key|Data associated with the given group key should be retracted
+
+### Memory Types
+
+Data memory references memory used for storing user data.
+This is in contrast to process or program memory.
+Since flux is integrated in co-tenant environments, it needs to handle arbitrary user data and avoid crashing the system or causing a noisy neighbor performance problem.
+Flux does this by separating the concept of process memory and data memory.
+
+Process memory is any memory required to execute code.
+Process memory is anything in Go that would allocate stack or heap memory and is tracked by the garbage collector.
+User code, by necessity, will use process memory.
+The amount of process memory used by a transformation should be designed to be fairly consistent.
+If a user has 10 rows or 10,000 rows, the process memory should not scale directly with the number of rows.
+This is not a strict requirement.
+There's no expectation that a transformation will use the same memory regardless of the number of rows, but just the requirement that it doesn't linearly scale with the number of rows or worse.
+
+Data memory has different conditions.
+Data memory is memory that is used to store user data and data memory is tracked by the `memory.Allocator`.
+The flux engine places limits on the amount of data memory that can be used and will abort execution if more is used.
+It is allowed and expected that some transformations will have bad memory footprints for certain inputs.
+
+The primary method of storing data is through immutable arrow arrays.
+There are also circumstances where mutable data is needed to implement an algorithm, but process memory is not an appropriate place to store that data (such as `distinct()` or `sort()`).
+See the **Mutable Data** section for how to handle these circumstances.
+
+## Execution Engine
+
+The execution engine is central to executing flux queries.
+
+### Pipeline Creation
+
+Before the execution engine begins, it's important to understand the steps that happen before execution.
+This is only a brief outline of those steps.
+
+* Text is parsed into an AST.
+* AST is converted into semantic graph.
+* Type inference runs on the semantic graph and assigns types to all nodes.
+* The interpreter uses a recursive-descent execution to evaluate the semantic nodes.
+* Side effects are gathered from the interpreter execution and side effects that are linked to table objects are collected into a spec.
```

Comment:
Might be good to define side-effects somewhere. This was always a point I was unclear on when learning flux. The side effects are actually what manifest as data destinations, be it `to()` or output downstream to the user. So the process of determining what the side effects are is actually determining the outputs of the entire pipeline. To me, It is different from the notion of side effects in a functional programming language, which is modification of some global state. So it's not enough to just say they are "side effects".